File: Task4A_B/task4a_b/src/main/java/com/cps2004/Ship.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class Ship extends Transport{
    private List<String> harbours;

    // Default constructor
    public Ship(){
        super();
        harbours = new ArrayList<>();
    }

    // Parameterized constructor
    public Ship(int transportSpeed, int costPerKm, int transportID, List<String> harbours){
        super(transportSpeed, costPerKm, transportID);
        this.harbours = harbours;
    }

    // Getters and setters
    public List<String> getRoute() { return harbours; }
    public void setHarbours(List<String> harbours) { this.harbours = harbours; }

    @Override
    // Calculate delivery plan
    public String deliveryPlan() {
        StringBuilder result = new StringBuilder("Ship will sail through: ");
        for (int i = 0; i < harbours.size(); i++) {
            if (i == harbours.size() - 1) {
                result.append(harbours.get(i));
            } else {
                result.append(harbours.get(i)).append(", ");
            }
        }
        return result.toString();
    }

}

File: Task1B/ship.cpp
#include "ship.h"

// Default constructor
Ship::Ship() {
    this->harbours = {};
}

// Parameterised constructor
Ship::Ship(int transportSpeed, int costPerKm, int id, vector<string> harbours) : Transport(transportSpeed, costPerKm, id) {
    this->harbours = harbours;
}


vector<string> Ship::getHarbours() {
    return this->harbours;
}

void Ship::setRoute(vector<string> harbours) {
    this->harbours = harbours;
}

// Display delivery plan
string Ship::deliveryPlan() {
    string result = "Ship will sail through: ";
    for (string harbour : this->getHarbours()) {
        result += harbour + ", ";
    }
    result = result.substr(0, result.length() - 2);
    return result;
}

// Save ship information
void Ship::save(ostream& out) const {
    out << "ship" << "\n";
    Transport::save(out);
    for (string motorway : this->harbours) {
        out << motorway << ",";
    }
    out << "\n";
}

// Load ship information
void Ship::load(istream& in) {
    Transport::load(in);
    string harboursStr;
    getline(in, harboursStr, '\n');
    stringstream ss(harboursStr);
    string harbour;
    harbours.clear();
    while (getline(ss, harbour, ',')) {
        harbours.push_back(harbour);
    }
}

// Destructor
Ship::~Ship() {
    this->harbours.clear();
}